eth/consensus/clique/clique.py

```diff
@@ -85,22 +87,23 @@ def __init__(self, base_db: AtomicDatabaseAPI, epoch_length: int = EPOCH_LENGTH)
             self._epoch_length,
         )
 
-    @to_tuple
-    def amend_vm_configuration(self, config: VMConfiguration) -> Iterable[VMFork]:
+    @classmethod
+    def amend_vm_configuration_for_chain_class(cls, config: VMConfiguration) -> None:
         """
         Amend the given ``VMConfiguration`` to operate under the rules of Clique consensus.
         """
         for pair in config:
             block_number, vm = pair
-            vm_class = vm.configure(
-                extra_data_max_bytes=65535,
-                validate_seal=staticmethod(self.validate_seal),
-                create_execution_context=staticmethod(self.create_execution_context),
-                configure_header=configure_header,
-                _assign_block_rewards=lambda _, __: None,
-            )
-
-            yield block_number, vm_class
+            setattr(vm, 'extra_data_max_bytes', 65535)
+            setattr(vm, 'create_execution_context', staticmethod(cls.create_execution_context))
+            setattr(vm, 'configure_header', configure_header)
+            setattr(vm, '_assign_block_rewards', lambda *_: None)
+
+    def amend_vm_for_chain_instance(self, vm: VirtualMachineAPI) -> None:
+        # `validate_seal` is stateful. We would not want two different instances of GoerliChain
+        # to operate on the same instance of CliqueConsensus. Therefore, this modification needs
+        # to be done for a specic chain instance, not chain class.
```

Comment:
Getting backed into a corner here, where this seems like the best solution, makes me think we're doing something else wrong at an architectural level.

One avenue to explore is the location of the consensus class. I'm thinking this might all get a lot easier if the `consensus_class` were a feature of the VM instead of the chain. Arguably, that better represents the real world anyway, as the consensus rules might change at a fork. Then the class gets initialized inside the VM. `VM.validate_seal()` could just forward to `VM.consensus_engine.validate_seal()`, etc.

Reply:
I was thinking the same especially because `validate_seal` is always going to be overwritten. The thing that I'm unsure about though: There's no clear picture about which other methods expect to be overwritten. Some consensus engine may fiddle with block rewards another might do funky things with uncles. If we take Clique as a driver then we end up formalizing `validate_seal `extra_data_max_bytes`, `create_execution_context`, `configure_header`, `_assign_block_rewards` for the consensus engine whereas `no_proof` and `pow` only need `validate_seal`. 

>Arguably, that better represents the real world anyway, as the consensus rules might change at a fork

But it also becomes really blurry where consensus rules go. I mean, every new VM is a consensus change and that's what VMs are for. Other consensus engines are a weird and barely formalized way of somehow hacking and bending the system (e.g. how `extra_data`, `nonce`, `coinbase` get totally different meaning in Clique) to somehow work. Something tells me that deserves a weird patchy answer and the current APIs at least break it down to a very generic level of hacking things on the class or the instance level.